wavefront/timeUniformSample.go

```diff
@@ -0,0 +1,222 @@
+	defer s.mutex.Unlock()
+	return len(s.rawValues())
+}
+
+// Snapshot returns a read-only copy of the sample.
+func (s *TimeUniformSample) Snapshot() metrics.Sample {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	values := make([]int64, s.values.len)
+	copy(values, s.rawValues())
+	return metrics.NewSampleSnapshot(s.values.len, values)
+}
+
+// StdDev returns the standard deviation of the values in the sample.
+func (s *TimeUniformSample) StdDev() float64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SampleStdDev(s.rawValues())
+}
+
+// Sum returns the sum of the values in the sample.
+func (s *TimeUniformSample) Sum() int64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SampleSum(s.rawValues())
+}
+
+// Update samples a new value.
+func (s *TimeUniformSample) Update(v int64) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.cleanOldValues()
+	if s.values.IsFull() {
+		s.values.Pop()
+	}
+	s.values.Push(&SampleValue{v: v, time: time.Now()})
+}
+
+func (s *TimeUniformSample) cleanOldValues() {
+	now := time.Now()
+	var needPop bool
+	for more := true; more; more = needPop {
+		sample, empty := s.values.Peek()
+		if empty {
+			needPop = false
+		} else {
+			needPop = (now.Sub(sample.(*SampleValue).time).Seconds() > s.lifeTime.Seconds())
+			if needPop {
+				s.values.Pop()
+			}
+		}
+	}
+}
+
+// Values returns a copy of the values in the sample.
+func (s *TimeUniformSample) Values() []int64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.rawValues()
+}
+
+func (s *TimeUniformSample) rawValues() []int64 {
+	values := make([]int64, s.values.len)
+	idx := 0
+	for _, value := range s.values.content {
+		if value != nil {
+			values[idx] = value.(*SampleValue).v
+			idx++
+		}
+	}
+	return values
+}
+
+// Variance returns the variance of the values in the sample.
+func (s *TimeUniformSample) Variance() float64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SampleVariance(s.rawValues())
+}
+
+type SampleValue struct {
+	v    int64
+	time time.Time
+}
+
+const MAX_QUEUE_SIZE = 1024
+
+type Queue struct {
+	content   [MAX_QUEUE_SIZE]interface{}
+	readHead  int
+	writeHead int
+	len       int64
+}
+
+func (q *Queue) IsFull() bool {
+	return q.len >= MAX_QUEUE_SIZE
+}
+
+func (q *Queue) Push(e interface{}) bool {
+	if q.len >= MAX_QUEUE_SIZE {
+		return false
+	}
+	q.content[q.writeHead] = e
+	q.writeHead = (q.writeHead + 1) % MAX_QUEUE_SIZE
+	q.len++
+	// fmt.Println("[QUEUE] - Push - q.len:", q.len, "(", MAX_QUEUE_SIZE, ")")
+	return true
+}
+
+func (q *Queue) Pop() (interface{}, bool) {
+	if q.len <= 0 {
+		return nil, false
+	}
+	result := q.content[q.readHead]
+	q.content[q.readHead] = nil
+	q.readHead = (q.readHead + 1) % MAX_QUEUE_SIZE
+	q.len--
+	// fmt.Println("[QUEUE] - Pop - q.len:", q.len, "(", MAX_QUEUE_SIZE, ")")
+	return result, true
+}
+
+func (q *Queue) Peek() (interface{}, bool) {
+	if q.len <= 0 {
+		return nil, true
+	}
+	result := q.content[q.readHead]
+	return result, false
+}
```

Comment:
Do we need the whole `Queue` structure here? I think a Go slice could do the job.

https://github.com/golang/go/wiki/SliceTricks
https://stackoverflow.com/a/26863706

Reply:
Queue structure is mostly for code readability...

Also, I'm not a Go expert, but `queue = queue[1:]` looks bad to me...
A comment on your stackoverflow link:

```
This is not a good approach, as every time queue[1:] is done, it moves the start of the slice to point to the next element, but does not release the spaced used by the dequeued element. Effectively it has unbounded memory growth in the slice storage. Additionally, if the queued element is a pointer, or a struct containing pointers, then the underlying slice storage will retain the memory to the dequeued element, causing a further memory leak. – tul Jun 8 at 10:11
```

but now we have memory metrics, so will try

Reply:
Let's retain the `Queue` struct then.